rabbitmq/main.go

```diff
@@ -0,0 +1,231 @@
+	}
+
+	data.queueName = os.Getenv("QUEUE_NAME")
+	if data.queueName == "" {
+		return data, fmt.Errorf("received empty queue name")
+	}
+
+	host := ""
+	if data.includeUnacked {
+		host = os.Getenv("HOST")
+	} else {
+		data.includeUnacked = true
+		host = os.Getenv("API_HOST")
+	}
+	if host == "" {
+		return data, fmt.Errorf("received empty host field")
+	}
+	data.url = host
+	return data, nil
+}
+
+func failOnError(logger *zap.Logger, err error, msg string) {
+	if err != nil {
+		logger.Fatal(msg, zap.Error(err))
+	}
+}
+
+func consumeMessage(data rabbitMQConnData, logger *zap.Logger, producerChannel, consumerChannel *amqp.Channel) {
+	msgs, err := consumerChannel.Consume(
+		data.queueName, // queue
+		"",             // consumer
+		false,          // auto-ack
+		false,          // exclusive
+		false,          // no-local
+		false,          // no-wait
+		nil,            // args
+	)
+	failOnError(logger, err, "failed while reading messages")
+
+	forever := make(chan bool)
+
+	go func() {
+		for d := range msgs {
+			success := handleFissionFunction(d, data, logger, producerChannel, consumerChannel)
+			if success {
+				d.Ack(false)
+			}
+		}
+	}()
+	logger.Info("Waiting for messages")
+	<-forever
+}
+
+func handleFissionFunction(d amqp.Delivery, data rabbitMQConnData, logger *zap.Logger, producerChannel, consumerChannel *amqp.Channel) bool {
+	var value string = string(d.Body)
+	// Generate the Headers
+	fissionHeaders := map[string]string{
+		"X-Fission-MQTrigger-Topic":      data.fissionMetadata.Topic,
+		"X-Fission-MQTrigger-RespTopic":  data.fissionMetadata.ResponseTopic,
+		"X-Fission-MQTrigger-ErrorTopic": data.fissionMetadata.ErrorTopic,
+		"Content-Type":                   data.fissionMetadata.ContentType,
+	}
+
+	// Create request
+	req, err := http.NewRequest("POST", data.fissionMetadata.FunctionURL, strings.NewReader(value))
+	if err != nil {
+		logger.Error("failed to create HTTP request to invoke function",
+			zap.Error(err),
+			zap.String("function_url", data.fissionMetadata.FunctionURL))
+		return false
+	}
+
+	for k, v := range fissionHeaders {
+		req.Header.Set(k, v)
+	}
+
+	// Make the request
+	var resp *http.Response
+	for attempt := 0; attempt <= data.fissionMetadata.MaxRetries; attempt++ {
+		// Make the request
+		resp, err = http.DefaultClient.Do(req)
+		if err != nil {
+			logger.Error("sending function invocation request failed",
+				zap.Error(err),
+				zap.String("function_url", data.fissionMetadata.FunctionURL),
+				zap.String("trigger", data.fissionMetadata.TriggerName))
+			continue
+		}
+		if resp == nil {
+			continue
+		}
+		if err == nil && resp.StatusCode == http.StatusOK {
+			// Success, quit retrying
+			break
+		}
+	}
+
+	if resp == nil {
+		logger.Warn("every function invocation retry failed; final retry gave empty response",
+			zap.String("function_url", data.fissionMetadata.FunctionURL),
+			zap.String("trigger", data.fissionMetadata.TriggerName))
+		return false
+	}
```

Comment:
I would say this is common to all types of triggers and can be abstracted out as a function?

Reply:
Done